File: dobot/controller/event.py
from . import BaseHandler
from ..apps import handle_event


class BaseEventsAPIHandler(BaseHandler):

    def post(self):
        body = json_decode(self.request.body)
        req_type = body.get('type', None)

        if not req_type:
            pass

        return handle_event(event=req_type, body=body)

File: controllers/event.js
let handler = {
    'events': {},

    'add': function(event, callback, force = false){
        if(event in this.events && !force){
            //TODO: throw error that will stop node
        }

        console.info(`Loading Event: \n\t${event} loaded\n`);

        this.events[event] = callback;

        return this;
    },

    'fire': async function(event, request, response){
        if(!(event in this.events)){
            response.status(404);
            response.send('The event: {} is not registed with do.bot.');
            return;
        }
        
        console.info(`Firing Event: ${event}\n`);

        return await this.events[event](request, response);
    }
};

const controller = {
    'post': async function(request, response){
        const body = request.body;

        return await handler.fire(body.event.type, request, response);
    }
};


module.exports = {
    'controller': controller,
    'handler': handler
};
